### PR TITLE
Allow macOS plugins to register as app delegates

### PR DESCRIPTION
**Pull request by stuartmorgan**
Adds `addApplicationDelegate:` to the macOS plugin registrar, following the corresponding iOS method, and wires it up to the existing app delegation forwarding that was recently added for use at the application level. (The actual delegate is non-trivially different between iOS and macOS, but that's not resolveable without a complex migration on the iOS side, so the APIs currently diverge after the level of the `addApplicationDelegate:` method itself.)

This doesn't add any new methods to the delegation; those will be added in a follow-up PR.

Also fixes a retain cycle in the termination handler that prevented the new test from working.

Most of https://github.com/flutter/flutter/issues/41471

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by stuartmorgan** [comment]
This is now tested and ready for review.

File: shell/platform/darwin/common/framework/Source/FlutterBinaryMessengerRelay.mm [hunk 1]
```diff
@@ -14,7 +14,7 @@ @implementation FlutterBinaryMessengerRelay
 - (instancetype)initWithParent:(NSObject<FlutterBinaryMessenger>*)parent {
   self = [super init];
   if (self != nil) {
-    self.parent = parent;
+    _parent = parent;
```
Comment (stuartmorgan) [inline shell/platform/darwin/common/framework/Source/FlutterBinaryMessengerRelay.mm:17]: This is unrelated, but as discussed in Discord I accidentally duplicated the recent work to adopt FlutterBinaryMessengerRelay on macOS as part of debugging my delloc test not working, and this was the only substantive diff between our versions. `init` methods should not call `self` (which we violate a lot unfortunately, but this one was trivial to fix), so I had done a drive-by fix to this as part of moving it.

I can revert if you'd rather.

File: shell/platform/darwin/macos/framework/Headers/FlutterAppDelegate.h [hunk 2]
```diff
@@ -10,6 +10,30 @@
...
 
+/**
+ * A protocol to be implemented by the `NSApplicationDelegate` of an application to enable the
+ * Flutter framework and any Flutter plugins to register to receive application life cycle events.
+ *
+ * Implementers should forward all of the `NSApplicationDelegate` methods corresponding to the
+ * handlers in FlutterAppLifecycleDelegate to any registered delegates.
+ */
+FLUTTER_DARWIN_EXPORT
+@protocol FlutterAppLifecycleProvider <NSObject>
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterAppDelegate.h:21]: This change isn't strictly necessary, but it makes us match iOS, and it makes the engine code that uses these methods cleaner.

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginMacOS.h [hunk 3]
```diff
@@ -22,7 +23,7 @@ NS_ASSUME_NONNULL_BEGIN
  * expand over time to more closely match the functionality of the iOS FlutterPlugin.
  */
 FLUTTER_DARWIN_EXPORT
-@protocol FlutterPlugin <NSObject>
+@protocol FlutterPlugin <NSObject, FlutterAppLifecycleDelegate>
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterPluginMacOS.h:26]: This is to match iOS. Personally I would have not done this, and had plugins that also wanted to be `FlutterAppLifecycleDelegate`s implement this themselves, but the ship has sailed for iOS and there's not really much downside, so I'm opting for consistency with iOS (as I've tried to do in plugin APIs in general, to minimize pain in code sharing in plugins).

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h [hunk 4]
```diff
@@ -52,6 +52,13 @@ FLUTTER_DARWIN_EXPORT
 - (void)addMethodCallDelegate:(nonnull id<FlutterPlugin>)delegate
                       channel:(nonnull FlutterMethodChannel*)channel;
 
+/**
+ * Registers the plugin as a receiver of `NSApplicationDelegate` calls.
+ *
+ * @param delegate The receiving object, such as the plugin's main class.
+ */
+- (void)addApplicationDelegate:(nonnull NSObject<FlutterAppLifecycleDelegate>*)delegate;
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h:60]: This is a slight deviation from iOS, which uses `FlutterPlugin` as the argument type, but that shouldn't prevent any code sharing.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 5]
```diff
@@ -165,7 +170,7 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result;
 #pragma mark -
 
 @implementation FlutterEngineTerminationHandler {
-  FlutterEngine* _engine;
+  __weak FlutterEngine* _engine;
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:173]: This fixes a retain cycle that will affect essentially all real-word, non-add-to-app cases (and was breaking my new test).

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 6]
```diff
@@ -448,12 +465,12 @@ - (instancetype)initWithName:(NSString*)labelPrefix
   [self setUpAccessibilityChannel];
   [self setUpNotificationCenterListeners];
   id<NSApplicationDelegate> appDelegate = [[NSApplication sharedApplication] delegate];
-  const SEL selector = @selector(addApplicationLifecycleDelegate:);
-  if ([appDelegate respondsToSelector:selector]) {
+  if ([appDelegate conformsToProtocol:@protocol(FlutterAppLifecycleProvider)]) {
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:468]: This is a cleaner way to express the check, and also is what the iOS engine does.

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm [hunk 7]
```diff
@@ -515,7 +572,7 @@ - (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication* _Nonnu
                ICUDataPath:[fixtures stringByAppendingString:@"/icudtl.dat"]];
     FlutterEngine* engine = [[FlutterEngine alloc] initWithName:@"test"
                                                         project:project
-                                         allowHeadlessExecution:true];
+                                         allowHeadlessExecution:YES];
```
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:575]: Drive-by, noticed while comparing your test to my non-working test (now fixed).

File: shell/platform/darwin/common/framework/Source/FlutterBinaryMessengerRelay.mm (hunk 1 above)
Comment (cbracken) [reply shell/platform/darwin/common/framework/Source/FlutterBinaryMessengerRelay.mm:17]: Ah good call. Thanks.

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginMacOS.h (hunk 3 above)
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Headers/FlutterPluginMacOS.h:26]: Agreed - getting the headers as similar as possible both for consistency and with an eye to a future world in which we can merge the non-UIKit/AppKit specific parts implies that we've got to live with the union of past decisions across both that we might prefer to not have made in retrospect.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm (hunk 5 above)
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:173]: Thanks for spotting this.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm (hunk 6 above)
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:468]: Yeah less raw selector wrangling is always a plus.

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm [hunk 8]
```diff
@@ -59,6 +63,59 @@ - (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication* _Nonnu
...
+  std::vector<void*> _delegates;
+}
+
+- (void)addApplicationLifecycleDelegate:(nonnull NSObject<FlutterAppLifecycleDelegate>*)delegate {
+  _delegates.push_back((__bridge void*)delegate);
+}
+
+- (void)removeApplicationLifecycleDelegate:
+    (nonnull NSObject<FlutterAppLifecycleDelegate>*)delegate {
+  auto i = std::find(_delegates.begin(), _delegates.end(), (__bridge void*)delegate);
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:95]: Maybe `delegateIndex` for the minor readability benefit given the type is `auto`?

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm (hunk 7 above)
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:575]: Ugh - thanks. I thought I'd done a sweep of these at one point but clearly missed a few. Will re-check today; maybe I never committed that branch?

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm (hunk 8 above)
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:95]: Done.
